org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/corrections/proposals/NewCUProposal.java

```diff
@@ -0,0 +1,441 @@
+		} else if (fTypeContainer instanceof IPackageFragment) {
+			String name = ASTNodes.getSimpleNameIdentifier(fNode);
+			ICompilationUnit parentCU = ((IPackageFragment) fTypeContainer).getCompilationUnit(getCompilationUnitName(name));
+			targetType = parentCU.getType(name);
+
+			CompositeChange change = new CompositeChange(fName);
+			change.add(new CreateFileChange(targetType.getResource().getRawLocation(), "", ""));
+			change.add(constructNewCUChange(parentCU));
+			return change;
+		} else {
+			return null;
+		}
+	}
+
+	private TextEdit constructEnclosingTypeEdit(ICompilationUnit icu) throws CoreException {
+		ASTParser astParser = ASTParser.newParser(IASTSharedValues.SHARED_AST_LEVEL);
+		astParser.setSource(icu);
+		CompilationUnit cu = (CompilationUnit) astParser.createAST(null);
+		TypeDeclaration enclosingDecl = findEnclosingTypeDeclaration(cu, fTypeContainer.getElementName());
+		AST ast = cu.getAST();
+		ASTRewrite rewrite = ASTRewrite.create(ast);
+		final AbstractTypeDeclaration newDeclaration;
+		switch (fTypeKind) {
+			case K_CLASS:
+				newDeclaration = ast.newTypeDeclaration();
+				((TypeDeclaration) newDeclaration).setInterface(false);
+				break;
+			case K_INTERFACE:
+				newDeclaration = ast.newTypeDeclaration();
+				((TypeDeclaration) newDeclaration).setInterface(true);
+				break;
+			case K_ENUM:
+				newDeclaration = ast.newEnumDeclaration();
+				break;
+			case K_ANNOTATION:
+				newDeclaration = ast.newAnnotationTypeDeclaration();
+				break;
+			default:
+				return null;
+		}
+		newDeclaration.setJavadoc(null);
+		newDeclaration.setName(ast.newSimpleName(ASTNodes.getSimpleNameIdentifier(fNode)));
+		newDeclaration.modifiers().add(ast.newModifier(Modifier.ModifierKeyword.PUBLIC_KEYWORD));
+		if (isParameterizedType(fTypeKind, fNode)) {
+			addTypeParameters((TypeDeclaration) newDeclaration);
+		}
+
+		ListRewrite lrw = rewrite.getListRewrite(enclosingDecl, TypeDeclaration.BODY_DECLARATIONS_PROPERTY);
+		lrw.insertLast(newDeclaration, null);
+		return rewrite.rewriteAST();
+	}
+
+	private void addTypeParameters(TypeDeclaration newDeclaration) {
+		if (isParameterizedType(fTypeKind, fNode)) {
+			String typeArgBaseName = getTypeArgBaseName(ASTNodes.getSimpleNameIdentifier(fNode));
+			int nTypeArgs = ((ParameterizedType) fNode.getParent().getParent()).typeArguments().size();
+			String[] typeArgNames = new String[nTypeArgs];
+			if (nTypeArgs == 1) {
+				typeArgNames[0] = typeArgBaseName;
+			} else {
+				for (int i = 0; i < nTypeArgs; i++) {
+					StringBuilder buf = new StringBuilder(typeArgBaseName);
+					buf.append(i + 1);
+					typeArgNames[i] = buf.toString();
+				}
+			}
+
+			AST ast = newDeclaration.getAST();
+			for (String typeArgName : typeArgNames) {
+				TypeParameter typeArg = ast.newTypeParameter();
+				typeArg.setName(ast.newSimpleName(typeArgName));
+				newDeclaration.typeParameters().add(typeArg);
+			}
+		}
+
+	}
+
+	private static String getTypeArgBaseName(String typeName) {
```

Comment:
This is for generic types, right? Rename to `getGenericTypeArgBaseName`.

Reply:
sure.